pkgs/tools/backup/sanoid/default.nix

```diff
@@ -0,0 +1,53 @@
+{ lib, stdenv, fetchFromGitHub, makeWrapper, coreutils, zfs, perlPackages
+, procps, which, openssh, sudo, mbuffer, pv, lzop, gzip, pigz }:
+
+with lib;
+
+stdenv.mkDerivation rec {
+  pname = "sanoid";
+  version = "2.0.2";
+
+  src = fetchFromGitHub {
+    owner = "jimsalterjrs";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "09cgchhpprr8yyx9kabwz3y7lz9kzn6wfdsqq3zam7c7yck672xa";
+  };
+
+  nativeBuildInputs = [ makeWrapper ];
+  buildInputs = with perlPackages; [ perl ConfigIniFiles CaptureTiny ];
+
+  installPhase = ''
+    mkdir -p "$out/bin"
+    mkdir -p "$out/etc/sanoid"
+    cp sanoid.defaults.conf "$out/etc/sanoid/sanoid.defaults.conf"
+    # Hardcode path to default config
+    substitute sanoid "$out/bin/sanoid" \
+      --replace "\$args{'configdir'}/sanoid.defaults.conf" "$out/etc/sanoid/sanoid.defaults.conf" \
+      --replace /sbin/zfs "${zfs}/bin/zfs" \
+      --replace /sbin/zpool "${zfs}/bin/zpool" \
```

Comment:
These should be calling the `/run/booted-system/sw/bin` equivalents to prevent kernel/user version mismatches. Unfortunately, that breaks the package for Nix-on-other-Linux systems.

Reply:
i believe there are partial patches for this upstream, which should be looked into.

Reply:
https://github.com/jimsalterjrs/sanoid/issues/393 https://github.com/jimsalterjrs/sanoid/pull/399 https://github.com/jimsalterjrs/sanoid/pull/399
There doesn't seem to be a patch for findoid yet.

Reply:
@deliciouslytyped Looking them up in `PATH` is very different from using those in `/run/booted-system/sw/bin`. Here's what could go wrong:

1. User upgrades system with `nixos-rebuild switch`, zfs changes from 0.8 to 0.9
2. 0.8 modules are still loaded, but 0.9 tools are now in `PATH`
3. User calls zfs/zpool, but there is some incompatibility between 0.8 and 0.9, and something breaks

There is no guarantee that mismatched versions will cooperate. Using `/run/booted-system/sw/bin` prevents this because it doesn't change on a `nixos-rebuild switch`, but instead has the versions the system booted with. This should ensure compatibility unless the user manually reloads kernel modules.

Reply:
Ah, I see. Shouldn't we change path to use that then though as opposed to poking the guts of upstream? I feel like this should be some FAQ entry instead of me asking in a dark corner of some issue. :D

Reply:
I applied the patches in https://github.com/jimsalterjrs/sanoid/pull/399 https://github.com/jimsalterjrs/sanoid/pull/451, and then added both `/run/booted-system/sw/bin` and `${zfs}/bin`  (in that order) to `PATH` using a wrapper. This will make sanoid use the correct version on NixOS, while still having a fallback for other OSes. I also put all the other dependencies first in `PATH` so they don't get pulled from `/run/booted-system/sw/bin`.

Reply:
I applied the patches in https://github.com/jimsalterjrs/sanoid/pull/399 https://github.com/jimsalterjrs/sanoid/pull/451, and then added both `/run/booted-system/sw/bin` and `${zfs}/bin`  (in that order) to `PATH` using a wrapper. This will make sanoid use the correct version on NixOS, while still having a fallback for other OSes. I also put all the other dependencies first in `PATH` so they don't get pulled from `/run/booted-system/sw/bin`.